File: core/Sources/ChotkiCore/Presentation/RecurrenceForm.swift
import Foundation

/// The editor's view of a recurrence, and the way back.
///
/// Extracted from the view so the round trip can be tested. It had three silent
/// data-loss bugs: a one-off day became a daily rule, a Great Lent rule became
/// a general fast-day rule, and a monthly rule's short-month policy reset. Each
/// happened because the form could not express the shape it had loaded, so
/// saving replaced it with something else — without saying so.
public struct RecurrenceForm: Equatable, Sendable {

    public enum Kind: String, CaseIterable, Hashable, Sendable {
        case once = "Just one day"
        case daily = "Every day"
        case weekly = "Certain weekdays"
        case monthly = "Once a month"
        case fastDays = "Fast days"
        case season = "Through a fasting season"
        case greatFeasts = "Great feasts"
    }

    public var kind: Kind = .daily
    public var weekdays: Set<Weekday> = [.sunday]
    public var monthDay: Int = 1
    /// No control for this; carried through so an edit cannot change it.
    public var shortMonthPolicy: ShortMonthPolicy = .lastDay
    public var season: FastingSeason = .greatLent
    public var onceDate: CalendarDate?

    public init() {}

    public init(_ recurrence: Recurrence) {
        switch recurrence {
        case .once(let day):
            kind = .once; onceDate = day
        case .daily:
            kind = .daily
        case .weekly(let days):
            kind = .weekly; weekdays = days
        case .monthly(let day, let policy):
            kind = .monthly; monthDay = day; shortMonthPolicy = policy
        case .liturgical(let trigger):
            switch trigger {
            case .fastDay: kind = .fastDays
            case .greatFeast: kind = .greatFeasts
            case .season(let which): kind = .season; season = which
            }
        }
    }

    /// `fallback` is used only when a one-off rule somehow has no day, so that
    /// it stays a single day rather than quietly becoming a daily rule.
    public func recurrence(fallback: CalendarDate) -> Recurrence {
        switch kind {
        case .once: return .once(onceDate ?? fallback)
        case .daily: return .daily
        case .weekly: return .weekly(days: weekdays.isEmpty ? [.sunday] : weekdays)
        case .monthly: return .monthly(day: monthDay, whenShort: shortMonthPolicy)
        case .fastDays: return .liturgical(.fastDay)
        case .season: return .liturgical(.season(season))
        case .greatFeasts: return .liturgical(.greatFeast)
        }
    }
}

File: android/core/src/main/kotlin/org/chotki/core/RecurrenceForm.kt
package org.chotki.core

/**
 * The editor's view of a recurrence, and the way back.
 *
 * Extracted from the view so the round trip can be tested. It had three silent
 * data-loss bugs: a one-off day became a daily rule, a Great Lent rule became a
 * general fast-day rule, and a monthly rule's short-month policy reset. Each
 * happened because the form could not express the shape it had loaded, so saving
 * replaced it with something else — without saying so.
 */
data class RecurrenceForm(
    val kind: Kind = Kind.DAILY,
    val weekdays: Set<Weekday> = setOf(Weekday.SUNDAY),
    val monthDay: Int = 1,
    /** No control for this; carried through so an edit cannot change it. */
    val shortMonthPolicy: ShortMonthPolicy = ShortMonthPolicy.LAST_DAY,
    val season: FastingSeason = FastingSeason.GREAT_LENT,
    val onceDate: CalendarDate? = null,
) {
    enum class Kind(val label: String) {
        ONCE("Just one day"),
        DAILY("Every day"),
        WEEKLY("Certain weekdays"),
        MONTHLY("Once a month"),
        FAST_DAYS("Fast days"),
        SEASON("Through a fasting season"),
        GREAT_FEASTS("Great feasts"),
    }

    companion object {
        fun of(recurrence: Recurrence): RecurrenceForm = when (recurrence) {
            is Recurrence.Once -> RecurrenceForm(kind = Kind.ONCE, onceDate = recurrence.date)
            Recurrence.Daily -> RecurrenceForm(kind = Kind.DAILY)
            is Recurrence.Weekly -> RecurrenceForm(kind = Kind.WEEKLY, weekdays = recurrence.days)
            is Recurrence.Monthly -> RecurrenceForm(
                kind = Kind.MONTHLY,
                monthDay = recurrence.day,
                shortMonthPolicy = recurrence.whenShort,
            )
            is Recurrence.Liturgical -> when (val trigger = recurrence.trigger) {
                LiturgicalTrigger.FastDay -> RecurrenceForm(kind = Kind.FAST_DAYS)
                LiturgicalTrigger.GreatFeast -> RecurrenceForm(kind = Kind.GREAT_FEASTS)
                is LiturgicalTrigger.Season ->
                    RecurrenceForm(kind = Kind.SEASON, season = trigger.season)
            }
        }
    }

    /**
     * [fallback] is used only when a one-off rule somehow has no day, so that it
     * stays a single day rather than quietly becoming a daily rule.
     */
    fun recurrence(fallback: CalendarDate): Recurrence = when (kind) {
        Kind.ONCE -> Recurrence.Once(onceDate ?: fallback)
        Kind.DAILY -> Recurrence.Daily
        Kind.WEEKLY -> Recurrence.Weekly(weekdays.ifEmpty { setOf(Weekday.SUNDAY) })
        Kind.MONTHLY -> Recurrence.Monthly(monthDay, shortMonthPolicy)
        Kind.FAST_DAYS -> Recurrence.Liturgical(LiturgicalTrigger.FastDay)
        Kind.SEASON -> Recurrence.Liturgical(LiturgicalTrigger.Season(season))
        Kind.GREAT_FEASTS -> Recurrence.Liturgical(LiturgicalTrigger.GreatFeast)
    }
}
